README.md
# Kiku

![Kiku](https://raw.githubusercontent.com/jrmykolyn/sfco-kiku/master/kiku.gif)

## Table of Contents
- [About](#about)
- [Features](#features)
- [Installation](#installation)
- [Setup](#setup)
- [Usage](#usage)
- [Documentation](#documentation)
- [Contributing](#contributing)
	- [Contibuting Overview](#contributing-overview)
	- [Code Style](#code-style)
	- [Testing](#testing)
- [Attribution](#attribution)


## About
Listen for keyboard input and invoke callbacks each time a specific string is encountered.

## Features
- Update the view, enable/disable bonus features, or just log data to the console when the user enters a string of your choosing.
- Listen for one or many strings (and their corresponding callbacks).

## Installation
```
npm install --save sfco-kiku
```

## Setup
Once installed, Kiku can be setup in the following ways:

- Copy the Kiku script from the `node_modules/` folder into the dependent project; inject the Kiku script into the document using a `<script>` tag.
- Pull Kiku into a 'vendor script' bundle and reference the resulting file.

## Usage
Once enqueued, the Kiku script exposes the `Kiku` constructor as a property of the `window` object. Create a new Kiku instance as follows:

```
let kikuRef = new Kiku();
```

Please note that there can be only 1x Kiku instance at any given time. For example, the following will fail:

```
let myFirstKikuRef = new Kiku();
let mySecondKikuRef = new Kiku(); // Error
```

The Kiku constructor accepts an optional `options` object which can be used to provide the initial strings to listen for, their callback functions, and various configuration data.

```
// Create a Kiku instance and register a callback for the string 'hello'.
let kikuRef = new Kiku( {
	bindings: [
		{
			string: 'hello',
			fn: () => { console.log( 'Hello, world' ); }
		}
	],
} );
```
```
// Create a Kiku instance and set the trigger key to 'space'.
let kikuRef = new Kiku( {
	settings: {
		triggerKey: 32, // 'Space'
	}
} );
```

Additional listeners and callback functions can be provided after instantiation using the `add()` method.

```
let kikuRef = new Kiku();

// Invoke `add()` with an object to add 1x new callback.
kikuRef.add( {
	string: 'foo',
	fn: () => { console.log( 'Bar' ); }
} );

// Invoke `add()` with an array of objects to add multiple callbacks.
kikuRef.add( [
	{
		string: 'baz',
		fn: () => { console.log( 'Quux' ); }
	},
	{
		string: 'beep',
		fn: () => { console.log( 'Boop' ); }
	}
] );
```

Existing listeners and their callbacks can be removed using the `remove()` method.

```
// Invoke `remove()` with a string to remove 1x callback.
kikuRef.remove( 'foo' );

// Invoke `remove()` with an array to remove multiple callbacks.
kikuRef.add( [ 'bar', 'beep' ] );
```

## API

### Kiku( options? )

Returns a Kiku instance which exposes:
- add() (Function)
- remove() (Function)

**options**

Type: `Object`

A wrapper around all data that Kiku can receive at instantiation time.

**options.settings**

Type: `Object`

An object of data that can be used to override default values.

**options.settings.caseSensitive**

Type: `boolean`

Default: `true`

Whether or not the user input should match the binding string exactly?

```
kikuRef.add( {
	string: 'hello',
	fn: function() {
		console.log( 'Hello, world!' );
	}
} );

When `caseSensitive` is set to `true`, the example above will only be triggered by:
- 'hello'

When `caseSensitive` is set to `false`, the example above will be triggered by:
- 'hello'
- 'HELLO'
- 'HeLlO'
- 'HELLo'
- etc.
```

**options.settings.triggerKey**

Type: `number`

Default: `32` (Enter)

Keycode for the key which is used to activate/deactive Kiku's listening functionality.

**options.settings.dismissKey**

Type: `number`

Default: `27` (Esc)

Keycode for the key which is used to cancel the listening functionality.

**options.bindings**

Type: `Array<Object>`

Default: `[]`

An array of object data. Each object describes a single string/callback pair, and must include the following keys:

- `string`: The string to listen for/check against.
- `fn`: The function to call if/when the `string` is matched.

## Documentation
Currently, Kiku *does not* include any external documentation.

For an overview of the project's evolution, please consult the `CHANGELOG`.

## Contributing

### Contributing Overview
Issues and proposed enhancements are welcome!

### Code Style
`ESlint` and `editorconfig` are used to enforce consistent code style and formatting. Please ensure that both of these tools are available within your IDE.

### Testing
Whoops, Kiku doesn't ship with any tests. Want to add some? Spin up an [issue](https://github.com/jrmykolyn/sfco-kiku/issues)!.

## Attribution
- `README.md` gif: https://giphy.com/gifs/halloween-ghost-ghosts-Yph6D7zPIVtIc


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## [0.2.0] - 2018-05-01
### Added
- Add support for programmatic activation/deactivation of Kiku instance via `activate()` and `deactivate()` methods. (topic-3)
- Add support for `getState()` instance method. (topic-5)

### Changed
- Update demo.

## [0.1.1] - 2018-04-24
### Added
- Add support for case sensitivity (active by default, see `README.md` for details). (topic-1)

### Changed
- Change primary event listener from `keyup` to `keydown`.

## [0.1.0]
### Added
- Everything.


gulpfile.js
// --------------------------------------------------
// IMPORT MODULES
// --------------------------------------------------
// Node

// Vendor
const eslint = require( 'gulp-eslint' );
const eslintConfigLocal = require( './.eslintrc.json' );
const gulp = require( 'gulp' );
const gulpIf = require( 'gulp-if' );
const babel = require( 'gulp-babel' );
const merge = require( 'deepmerge' );
const PathMap = require( 'sfco-path-map' );
const sjmConfigs = require( '@sjmdev/sjm-configs' );

// --------------------------------------------------
// DECLARE VARS
// --------------------------------------------------
const { eslintConfig } =  sjmConfigs;
const PATHS = new PathMap( {
	'src': 'src',
	'dest': 'dist',
} );

// --------------------------------------------------
// DEFINE TASKS
// --------------------------------------------------
/**
 * Wrapper around any/all tasks to be executed when `gulp` is run.
 */
gulp.task( 'default', [ 'scripts' ] );

gulp.task( 'scripts:lint', () => {
	return gulp.src( `${PATHS.src}/*.js` )
		.pipe( eslint( merge( eslintConfig, eslintConfigLocal, {
			arrayMerge: ( a, b ) => { return b; },
		} ) ) )
		.pipe( eslint.format() );
} );

gulp.task( 'scripts:fix', () => {
	return gulp.src( `${PATHS.src}/*.js` )
		.pipe( eslint( {
			rules: merge( eslintConfig.rules, eslintConfigLocal.rules, {
				arrayMerge: ( a, b ) => { return b; },
			} ),
			fix: true,
		} ) )
		.pipe( eslint.format() )
		.pipe( gulpIf(
			( file ) => {
				return !!file.eslint && file.eslint.fixed;
			},
			gulp.dest( PATHS.src )
		) );
} );

gulp.task( 'scripts:compile', function() {
	return gulp.src( `${PATHS.src}/*.js` )
		.pipe( babel( {
			presets: [ 'env' ],
			plugins: [
				[
					'transform-es2015-modules-umd',
					{
						globals: {
							'kiku': 'Kiku',
						},
						exactGlobals: true,
					}
				],
			],
		} ) )
		.pipe( gulp.dest( PATHS.dest ) );
} );

/**
 * Wrapper around script-related tasks.
 *
 * NOTE: `scripts:fix` task is exluded (ie. it must be run manually).
 */
gulp.task( 'scripts', [ 'scripts:lint', 'scripts:compile' ] );


js/kiku.0.0.0.js
/**
 * Kiku is a plugin which allows for functions to
 * be invoked in response to user-specified strings
 * of alphanumeric characters.
 *
 * Kiku follows the 'singleton' pattern, meaning that
 * only one instance may exist at a given time.
 *
 * The Kiku instance exposes a public API which allows
 * the invoking script to view, access, and update
 * the instance's settings and data.
 *
 * @summary   Kiku is a plugin which allows for functions
 *			  to be invoked in response to user-specified
 * 			  strings of alphanumeric characters.
 *
 * @link      N/A
 * @since     0.0.0 (if available)
 * @requires  N/A
 *
 * @author    Jesse R Mykolyn <jrmykolyn@gmail.com>
 */

(function(global, document) {
	/* ---------------------------------------------------------------------------------------------------- */
	/* Private Vars */
	/* ---------------------------------------------------------------------------------------------------- */
	/**
	 * Stores a reference to the 'singleton' Kiku instance.
	*/
	var _self = {},
		_defaults = {
			data: {
				bindings: [],
				functions: {}
			},
			settings: {
				triggerKey: 13
			}
		};

	/* ---------------------------------------------------------------------------------------------------- */
	/* Private Functions */
	/* ---------------------------------------------------------------------------------------------------- */
	/**
	 * Initialize and return the public API for the Kiku Instance
	 *
	 * @param {Object} `options`
	 * @return {Object}
	*/
	function init(options) {		
		// Initialize Kiku instance
		_self.init = true;

		// Re-assign `options` object or fallack to empty obj.
		options = (options instanceof Object) ? options : {};

		// Validate/update instance settings && data.
		_self.settings = validateSettings(options.settings || null);
		_self.data = validateData(options.data || null);

		// Reconfigure functions and add global event listeners.
		parseFunctionBindings(_self);
		addEventListeners(global, _self);

		// Expose public API
		return {
			logSettings: function() {
				var output = [],
					str = '';

				for (var key in _self.settings) {
					str += (key + ': ' + _self.settings[key]);

					output.push(str);

					str = '';
				}

				return output;
			},
			getFunctionKeys: function() {
				return Object.keys(_self.data.functions);
			}
		};
	}


	/**
	 * Prints messages to the console relating to the invalid creation of a new Kiku instance.
	 *
	 * @param {Object} `context`
	*/
	function handleInvalidInstantiation(context) {
		var method = (typeof console.error !== 'undefined') ? 'error' : 'log';

		if (context === global) {
			console[method]('`Kiku` MUST BE INITIALIZED USING THE `new` KEYWORD'); // TEMP
		} else {
			console[method]('`Kiku` HAS ALREADY BEEN INITIALIZED'); // TEMP	
		}
	}


	/**
	 * Validates any settings received on instantiation.
	 * Modifies the `settings` argument to ensure that all required settings are present.
	 * Falls back to 'default settings' if argument is missing/invalid.
	 *
	 * @param {Object} `settings`
	 * @return {Object}
	*/
	function validateSettings(settings) {
		if (settings instanceof Object) {

			for (var key in _defaults.settings) {
				do {
					if (settings[key] && typeof settings[key] !== typeof _defaults.settings[key]) {
						settings[key] = _defaults.settings[key];
					}

					if (!settings[key]) {
						settings[key] = _defaults.settings[key];
					}
				} while (0);
			}

			return settings;
		} else {

			return _defaults.settings;
		}
	}


	/**
	 * Validates `data` - functions and corresponding strings - received on instantiation.
	 * Modifies the `data` argument to ensure that all required keys are present.
	 * Falls back to 'default data' if argument is missing/invalid.
	 *
	 * @param {Object} `data`
	 * @return {Object}
	*/
	function validateData(data) {
		if (data instanceof Object) {

			for (var key in _defaults.data) {
				do {
					if (data[key] && typeof data[key] !== typeof _defaults.data[key]) {
						data[key] = _defaults.data[key];
					}

					if (!data[key]) {
						data[key] = _defaults.data[key];
					}
				} while (0);
			}

			return data;
		} else {
		return _defaults.data;
		}
	}


	/**
	 * Loops over functions and strings received by Kiku instance, adds to property on self if valid.
	 *
	 * @param {Object} `selfObj`
	 * @return {'undefined'}
	*/
	function parseFunctionBindings(selfObj) {
		if (Array.isArray(selfObj.data.bindings)) {
			for (var i = 0, x = selfObj.data.bindings.length; i < x; i++) {
				var binding = selfObj.data.bindings[i];
				
				if (binding instanceof Object) {
					if (typeof binding.string === 'string' && binding.fn instanceof Function) {
						_self.data.functions[binding.string.toLowerCase()] = binding.fn;
					}
				}
			}
		}
	}


	/**
	 * Adds a 'keyup' listener to the `window` object.
	 * If event keycode is equal to the value of `triggerKey` (stored on the Kiku instance), `evaluateInput()` is invoked.
	 * Otherwise, `appendCharToInput()` is invoked.
	 *
	 * @param {Object} `context`
	 * @param {Object} `selfObj`
	*/
	function addEventListeners(context, selfObj) {
		context.addEventListener('keyup', function(e) {
			if (parseInt(e.keyCode) === selfObj.settings.triggerKey) {
				evaluateInput();
			} else {
				appendCharToInput(getCharFromKeyCode(parseInt(e.keyCode)));
			}
		});
	}


	/**
	 * Checks for function that matches the value of `input` property on the Kiku instance. Invokes function if valid.
	 * Resets the `input` property to an empty string.
	*/
	function evaluateInput() {
		if (_self.data.input) {
			var fn_key = _self.data.input.toLowerCase();

			if (_self.data.functions[fn_key] instanceof Function) {

				_self.data.functions[fn_key]();
			}

			_self.data.input = '';
		}
	}


	/**
	 * Adds the received `char` to the `input` property on the Kiku instance.
	 * Sets the `input` property to an empty string if it doesn't exist.
	 *
	 * @param {String} `char`
	*/
	function appendCharToInput(char) {
		if (!_self.data.input) { _self.data.input = ''; }

		_self.data.input += char;
	}


	/**
	 * Returns the alphabetical character for a given integer. 
	 *
	 * @param {Number} `keyCode`
	 * @return {String}
	*/
	function getCharFromKeyCode(keyCode) {
		return String.fromCharCode(keyCode);
	}


	/* ---------------------------------------------------------------------------------------------------- */
	/* Constructor */
	/* ---------------------------------------------------------------------------------------------------- */
	global.Kiku = function(options) {
		var _this = this;

		// If Kiku has not been instantiated and context is *not* `window`. 
		if (!_self.init && _this !== global) {
			return init(options);

		// Otherwise, display the appropriate error message(s).
		} else {
			handleInvalidInstantiation(_this);
		}
	};
})(window, document);

js/sfco-kiku.0.0.0.js
/**
 * Kiku is a JavaScript plugin which allows for functions
 * to be invoked in response to user-specified strings
 * of alphanumeric characters.
 *
 * Kiku follows the 'singleton' pattern, meaning that
 * only one instance may exist at a given time.
 *
 * The Kiku instance exposes a public API which allows
 * the invoking script to view, access, and update
 * the instance's settings and data.
 *
 * @summary   Kiku is a JavaScript plugin which allows for
 *			  functions to be invoked in response to
 *			  user-specified strings of alphanumeric
 *			  characters.
 *
 * @link      N/A
 * @since     0.0.0 (if available)
 * @requires  N/A
 *
 * @author    Jesse R Mykolyn <jrmykolyn@gmail.com>
 */

(function(global, document) {
	/* ---------------------------------------------------------------------------------------------------- */
	/* Private Vars */
	/* ---------------------------------------------------------------------------------------------------- */
	/**
	 * Stores a reference to the 'singleton' Kiku instance.
	*/
	var _self = {
		state: {
			isActive: false,
		},
	};

	var _defaults = {
			data: {
				bindings: [],
				functions: {}
			},
			settings: {
				triggerKey: 13,
				dismissKey: 27,
			},
		};

	/* ---------------------------------------------------------------------------------------------------- */
	/* Private Functions */
	/* ---------------------------------------------------------------------------------------------------- */
	/**
	 * Initialize and return the public API for the Kiku Instance
	 *
	 * @param {Object} `options`
	 * @return {Object}
	*/
	function init(options) {
		// Initialize Kiku instance
		_self.init = true;

		// Re-assign `options` object or fallack to empty obj.
		options = (options instanceof Object) ? options : {};

		// Validate/update instance settings && data.
		_self.settings = validateSettings(options.settings || null);
		_self.data = validateData(options.data || null);

		// Reconfigure functions and add global event listeners.
		parseFunctionBindings(_self);
		addEventListeners(global, _self);

		// Expose public API
		return {
			logSettings: function() {
				var output = [],
					str = '';

				for (var key in _self.settings) {
					str += (key + ': ' + _self.settings[key]);

					output.push(str);

					str = '';
				}

				return output;
			},
			getFunctionKeys: function() {
				return Object.keys(_self.data.functions);
			}
		};
	}

	/**
	 * Prints messages to the console relating to the invalid creation of a new Kiku instance.
	 *
	 * @param {Object} `context`
	*/
	function handleInvalidInstantiation(context) {
		var method = (typeof console.error !== 'undefined') ? 'error' : 'log';

		if (context === global) {
			console[method]('`Kiku` MUST BE INITIALIZED USING THE `new` KEYWORD'); // TEMP
		} else {
			console[method]('`Kiku` HAS ALREADY BEEN INITIALIZED'); // TEMP
		}
	}

	/**
	 * Validates any settings received on instantiation.
	 * Modifies the `settings` argument to ensure that all required settings are present.
	 * Falls back to 'default settings' if argument is missing/invalid.
	 *
	 * @param {Object} `settings`
	 * @return {Object}
	*/
	function validateSettings(settings) {
		if (settings instanceof Object) {

			for (var key in _defaults.settings) {
				do {
					if (settings[key] && typeof settings[key] !== typeof _defaults.settings[key]) {
						settings[key] = _defaults.settings[key];
					}

					if (!settings[key]) {
						settings[key] = _defaults.settings[key];
					}
				} while (0);
			}

			return settings;
		} else {

			return _defaults.settings;
		}
	}

	/**
	 * Validates `data` - functions and corresponding strings - received on instantiation.
	 * Modifies the `data` argument to ensure that all required keys are present.
	 * Falls back to 'default data' if argument is missing/invalid.
	 *
	 * @param {Object} `data`
	 * @return {Object}
	*/
	function validateData(data) {
		if (data instanceof Object) {

			for (var key in _defaults.data) {
				do {
					if (data[key] && typeof data[key] !== typeof _defaults.data[key]) {
						data[key] = _defaults.data[key];
					}

					if (!data[key]) {
						data[key] = _defaults.data[key];
					}
				} while (0);
			}

			return data;
		} else {
			return _defaults.data;
		}
	}

	/**
	 * Loops over functions and strings received by Kiku instance, adds to property on self if valid.
	 *
	 * @param {Object} `selfObj`
	 * @return {'undefined'}
	*/
	function parseFunctionBindings(selfObj) {
		if (Array.isArray(selfObj.data.bindings)) {
			for (var i = 0, x = selfObj.data.bindings.length; i < x; i++) {
				var binding = selfObj.data.bindings[i];

				if (binding instanceof Object) {
					if (typeof binding.string === 'string' && binding.fn instanceof Function) {
						_self.data.functions[binding.string.toLowerCase()] = binding.fn;
					}
				}
			}
		}
	}

	/**
	 * Adds a 'keyup' listener to the `window` object.
	 * If event keycode is equal to the value of `triggerKey` (stored on the Kiku instance), `evaluateInput()` is invoked.
	 * Otherwise, `appendCharToInput()` is invoked.
	 *
	 * @param {Object} `context`
	 * @param {Object} `selfObj`
	*/
	function addEventListeners(context, selfObj) {
		// Register core event listeners.
		context.addEventListener('keyup', function( e ) {
			var k = parseInt( e.keyCode );;

			// Handle cases where Kiku is active.
			if ( selfObj.state.isActive ) {
				switch ( k ) {
					case selfObj.settings.dismissKey:
						var e = new CustomEvent( 'KIKU_DISMISS' );
						window.dispatchEvent( e );
						break;
					case selfObj.settings.triggerKey:
						var e = new CustomEvent( 'KIKU_EVALUATE' );
						window.dispatchEvent( e );
						break;
					default:
						var e = new CustomEvent( 'KIKU_APPEND', { detail: { data: e } } );
						window.dispatchEvent( e );
						console.log( selfObj.data.input );
				}
			// Handle cases where Kiku is inactive.
			} else {
					switch ( k ) {
						case selfObj.settings.triggerKey:
							var e = new CustomEvent( 'KIKU_ACTIVATE' );
							window.dispatchEvent( e );
					}
			}

		});

		// Register custom event/Kiku-specific event listeners.
		context.addEventListener( 'KIKU_ACTIVATE', function( e ) {
			_self.state.isActive = true;
		} );

		context.addEventListener( 'KIKU_EVALUATE', function( e ) {
			evaluateInput();
			_self.state.isActive = false;
		} );

		context.addEventListener( 'KIKU_APPEND', function( e ) {
			appendCharToInput( getCharFromKeyCode( parseInt( e.detail.data.keyCode ) ) );
		} );

		context.addEventListener( 'KIKU_DISMISS', function( e ) {
			_self.state.isActive = false;
			/// TODO: Clear `data`.
		} );

		context.addEventListener( 'KIKU_ON_SUCCESS', function( e ) {
			/// TODO
		} );

		context.addEventListener( 'KIKU_ON_FAIL', function( e ) {
			/// TODO
		} );
	}

	/**
	 * Checks for function that matches the value of `input` property on the Kiku instance. Invokes function if valid.
	 * Resets the `input` property to an empty string.
	*/
	function evaluateInput() {
		if (_self.data.input) {
			var k = _self.data.input.toLowerCase();

			if (_self.data.functions[k] instanceof Function) {
				_self.data.functions[k]();

				window.dispatchEvent( ( new CustomEvent( 'KIKU_ON_SUCCESS' ) ) );
			} else {
				window.dispatchEvent( ( new CustomEvent( 'KIKU_ON_FAIL' ) ) );
			}

			_self.data.input = '';
		}
	}

	/**
	 * Adds the received `char` to the `input` property on the Kiku instance.
	 * Sets the `input` property to an empty string if it doesn't exist.
	 *
	 * @param {String} `char`
	*/
	function appendCharToInput(char) {
		if (!_self.data.input) { _self.data.input = ''; }

		_self.data.input += char;
	}

	/**
	 * Returns the alphabetical character for a given integer.
	 *
	 * @param {Number} `keyCode`
	 * @return {String}
	*/
	function getCharFromKeyCode(keyCode) {
		return String.fromCharCode(keyCode);
	}

	/* ---------------------------------------------------------------------------------------------------- */
	/* Constructor */
	/* ---------------------------------------------------------------------------------------------------- */
	global.Kiku = function(options) {
		var _this = this;

		// If Kiku has not been instantiated and context is *not* `window`.
		if (!_self.init && _this !== global) {
			return init(options);

		// Otherwise, display the appropriate error message(s).
		} else {
			handleInvalidInstantiation(_this);
		}
	};
})(window, document);


src/kiku.js
/* eslint-disable no-console */

/**
 * Kiku is a JavaScript plugin which allows for functions
 * to be invoked in response to user-specified strings
 * of alphanumeric characters.
 *
 * Kiku follows the 'singleton' pattern, meaning that
 * only one instance may exist at a given time.
 *
 * The Kiku instance exposes a public API which allows
 * the invoking script to view, access, and update
 * the instance's settings and data.
 *
 * @summary   Kiku is a JavaScript plugin which allows for
 *			  functions to be invoked in response to
 *			  user-specified strings of alphanumeric
 *			  characters.
 *
 * @link      N/A
 * @since     0.0.0 (if available)
 * @requires  N/A
 *
 * @author    Jesse R Mykolyn <jrmykolyn@gmail.com>
 */

// --------------------------------------------------
// Utilities
// --------------------------------------------------
/**
 * Clone an object.
 *
 * @param {Object} obj
 * @return {Object}
 */
const clone = ( obj ) => {
	// If the input is a not an object, return it.
	if ( typeof obj !== 'object' ) {
		return obj;
	}

	let output = {};

	// Iterate over keys of input object; clone nested objects; allow primitives to pass through.
	for ( let k in obj ) {
		output[ k ] = ( typeof obj[ k ] === 'object' ) ? clone( obj[ k ] ) : obj[ k ];
	}

	return output;
};

const utils = {
	clone,
};

// --------------------------------------------------
// Private Vars
// --------------------------------------------------
/**
 * Stores a reference to the 'singleton' Kiku instance.
 */
const _self = {
	state: {
		isActive: false,
		input: '',
	},
	defaults: {
		triggerKey: 13, // Enter
		dismissKey: 27, // Esc
		caseSensitive:  true,
	},
	settings: {},
	bindings: [],
};

/// TODO: Consider migrating into `_self` obj.
const keyCodeBlacklist = [
	16, // Shift
];

// --------------------------------------------------
// Private Functions
// --------------------------------------------------
/**
 * Prints messages to the console relating to the invalid creation of a new Kiku instance.
 *
 * @param {Object} `context`
 */
const handleInvalidInstantiation = ( context ) => {
	let method = ( typeof console.error !== 'undefined' ) ? 'error' : 'log';

	if ( context === window ) {
		console[ method ]( '`Kiku` MUST BE INITIALIZED USING THE `new` KEYWORD' ); // TEMP
	} else {
		console[ method ]( '`Kiku` HAS ALREADY BEEN INITIALIZED' ); // TEMP
	}
};

/**
 * Validates any data received on instantiation.
 *
 * Modifies the `data` object to ensure that all required keys are present.
 *
 * Falls back to default value if a given input is missing/invalid.
 *
 * @param {Object} `data`
 * @param {string} `key`
 * @return {Object}
 */
const validateInput = ( data, key ) => {
	// If the input is missing/invalid, return the entire default object.
	if ( !data || typeof data !== 'object' ) {
		return _self[ key ];
	}

	// Otherwise, migrate indiv. values from default object if required.
	for ( let k in _self[ key ] ) {
		if ( typeof data[ k ] === 'undefined' || typeof data[ k ] !== typeof _self[ key ][ k ] ) {
			data[ k ] = _self[ key ][ k ];
		}
	}

	return data;
};

/**
 * Registers required event listeners and handles input.
 *
 * @param {Object} `context`
 * @param {Object} `selfObj`
 */
const addEventListeners = ( context, selfObj ) => {
	// Register core event listeners.
	context.addEventListener( 'keydown', ( e ) => {
		let k = parseInt( e.keyCode );
		let customEvent;

		// Handle cases where Kiku is active.
		if ( selfObj.state.isActive ) {

			// Bail early if key is blacklisted.
			if ( keyCodeBlacklist.indexOf( k ) !== -1 ) {
				return;
			}

			switch ( k ) {
			case selfObj.settings.dismissKey:
				customEvent = new CustomEvent( 'KIKU_DISMISS' );
				window.dispatchEvent( customEvent );
				break;
			case selfObj.settings.triggerKey:
				customEvent = new CustomEvent( 'KIKU_EVALUATE' );
				window.dispatchEvent( customEvent );
				break;
			default:
				customEvent = new CustomEvent( 'KIKU_APPEND', { detail: { data: e } } );
				window.dispatchEvent( customEvent );
			}
		// Handle cases where Kiku is inactive.
		} else {
			switch ( k ) {
			case selfObj.settings.triggerKey:
				customEvent = new CustomEvent( 'KIKU_ACTIVATE' );
				window.dispatchEvent( customEvent );
			}
		}

	} );

	// Register custom event/Kiku-specific event listeners.
	context.addEventListener( 'KIKU_ACTIVATE', () => {
		_self.state.isActive = true;
	} );

	context.addEventListener( 'KIKU_EVALUATE', () => {
		evaluateInput();
		_self.state.isActive = false;
	} );

	context.addEventListener( 'KIKU_APPEND', ( e ) => {
		appendCharToInput( parseEvent( e.detail.data ) );
	} );

	context.addEventListener( 'KIKU_DISMISS', () => {
		_self.state.isActive = false;
		_self.state.data = '';
	} );

	context.addEventListener( 'KIKU_ON_SUCCESS', () => {
		/// TODO
	} );

	context.addEventListener( 'KIKU_ON_FAIL', () => {
		/// TODO
	} );
};

/**
 * Validate input and invoke corresponding function.
 */
const evaluateInput = () => {
	if ( _self.state.input ) {
		let str = ( !_self.settings.caseSensitive ) ? _self.state.input.toLowerCase() : _self.state.input;

		// Get `binding` object.
		let binding = _self.bindings.filter( ( binding ) => {
			let string = ( !_self.settings.caseSensitive ) ? binding.string.toLowerCase() : binding.string;
			return string === str;
		} )[ 0 ];

		// Validate, invoke, and dispatch event(s).
		if ( binding && typeof binding === 'object' && binding.fn instanceof Function ) {
			binding.fn();

			window.dispatchEvent( ( new CustomEvent( 'KIKU_ON_SUCCESS' ) ) );
		} else {
			window.dispatchEvent( ( new CustomEvent( 'KIKU_ON_FAIL' ) ) );
		}

		_self.state.input = '';
	}
};

/**
 * Adds the received `char` to the `input` property on the Kiku instance.
 * Sets the `input` property to an empty string if it doesn't exist.
 *
 * @param {String} `char`
 */
const appendCharToInput = ( char ) => {
	if ( !_self.state.input ) { _self.state.input = ''; }

	_self.state.input += char;
};

/**
 * Allow events which contain 1x alphanumeric key to pass through; parse complex keys.
 *
 * @param {Object} e
 * @return {string}
 */
const parseEvent = ( e ) => {
	let p = /^[A-z0-9]$/gmi;
	let output = ( e.key.length === 1 && p.test( e.key ) ) ? e.key : getCharFromKeyCode( e.keyCode );
	return output;
};

/**
 * Returns the alphabetical character for a given integer.
 *
 * @param {Number} `keyCode`
 * @return {String}
 */
const getCharFromKeyCode = ( keyCode ) => {
	return String.fromCharCode( keyCode );
};


// --------------------------------------------------
// Public Methods
// --------------------------------------------------
/**
 * Register a new string and callback function.
 *
 * @param {Array<Object>|Object} arr
 * @return {boolean}
 */
const add = ( arr ) => {
	// Ensure array.
	arr = ( Array.isArray( arr ) ) ? arr : [ arr ];

	// Validate and filter out duds.
	let vals = arr.filter( ( o ) => {
		return ( o && typeof o === 'object' && typeof o.string === 'string' && typeof o.fn === 'function' );
	} );

	// Update bindings.
	_self.bindings = ( vals.length ) ? [ ..._self.bindings, ...vals ] : _self.bindings;

	// Return value based on whether any new listeners/callbacks were added.
	return !!vals.length;
};

/**
 * Remove existing string and callback function.
 *
 * @param {Array<string>|string} arr
 * @return {boolean}
 */
const remove = ( arr ) => {
	// Ensure array.
	arr = ( Array.isArray( arr ) ) ? arr : [ arr ];

	// Validate.
	let vals = arr.filter( ( str ) => typeof str === 'string' );

	// Reassign bindings.
	let initLength = _self.bindings.length;
	_self.bindings = _self.bindings.filter( ( o ) => vals.indexOf( o.string ) === -1 );

	// Return value based on whether any listeners/callbacks were removed.
	return ( _self.bindings.length !== initLength );
};

/**
 * Activate Kiku instance.
 *
 * @return boolean
 */
const activate = () => {
	return ( _self.state.isActive = true );
};

/**
 * Deactivate Kiku instance.
 *
 * @return boolean
 */
const deactivate = () => {
	return ( _self.state.isActive = false );
};

/**
 * Return a copy of the Kiku instance's internal state.
 *
 * @return {Object}
 */
const getState = () => {
	return utils.clone( _self.state );
};

/**
 * Get a list of registered callback functions.
 *
 * @return {Array<string>}
 */
const getFunctionKeys = () => {
	return _self.bindings.map( binding => binding.string );
};

/**
 * Initialize and return the public API for the Kiku Instance
 *
 * @param {Object} `options`
 * @return {Object}
 */
const init = ( options ) => {
	// Initialize Kiku instance
	_self.init = true;

	// Re-assign `options` object or fallack to empty obj.
	options = ( options instanceof Object ) ? options : {};

	// Validate/update instance settings.
	_self.settings = validateInput( options.settings, 'defaults' );

	addEventListeners( window, _self );

	// Expose public API
	return {
		// Core
		add,
		remove,
		activate,
		deactivate,
		// Supporting
		getState,
		// getFunctionKeys, /// TODO: Consider exposing.
	};
};

// --------------------------------------------------
// Public API
// --------------------------------------------------
const Kiku = ( options ) => {
	let _this = this;

	// If Kiku has not been instantiated and context is *not* `window`.
	if ( !_self.init && _this !== window ) {
		return init( options );

	// Otherwise, display the appropriate error message(s).
	} else {
		handleInvalidInstantiation( _this );
	}
};

/* eslint-disable-next-line no-undef */
module.exports = Kiku;
